network/retrieve/retrieve.go

```diff
@@ -0,0 +1,429 @@
+		return
+	}
+	osp.LogFields(olog.Bool("delivered", true))
+}
+
+// handleChunkDelivery handles a ChunkDelivery message from a certain peer
+// if the chunk proximity order in relation to our base address is within depth
+// we treat the chunk as a chunk received in syncing
+func (r *Retrieval) handleChunkDelivery(ctx context.Context, p *Peer, msg *ChunkDelivery) error {
+	p.logger.Debug("retrieval.handleChunkDelivery", "ref", msg.Addr)
+	var osp opentracing.Span
+	ctx, osp = spancontext.StartSpan(
+		ctx,
+		"handle.chunk.delivery")
+
+	processReceivedChunksCount.Inc(1)
+
+	// record the last time we received a chunk delivery message
+	lastReceivedRetrieveChunksMsg.Update(time.Now().UnixNano())
+
+	// count how many chunks we receive for retrieve requests per peer
+	peermetric := fmt.Sprintf("chunk.delivery.%x", p.BzzAddr.Over()[:16])
+	metrics.GetOrRegisterCounter(peermetric, nil).Inc(1)
+
+	peerPO := chunk.Proximity(p.BzzAddr.Over(), msg.Addr)
+	po := chunk.Proximity(r.kad.BaseAddr(), msg.Addr)
+	depth := r.kad.NeighbourhoodDepth()
+	var mode chunk.ModePut
+	// chunks within the area of responsibility should always sync
+	// https://github.com/ethersphere/go-ethereum/pull/1282#discussion_r269406125
+	if po >= depth || peerPO < po {
+		mode = chunk.ModePutSync
+	} else {
+		// do not sync if peer that is sending us a chunk is closer to the chunk then we are
+		mode = chunk.ModePutRequest
+	}
+
+	p.logger.Trace("handle.chunk.delivery", "ref", msg.Addr)
+
+	go func() {
+		defer osp.Finish()
+		p.logger.Trace("handle.chunk.delivery", "put", msg.Addr)
+		_, err := r.netStore.Put(ctx, mode, storage.NewChunk(msg.Addr, msg.SData))
+		if err != nil {
+			if err == storage.ErrChunkInvalid {
+				p.Drop()
+			}
+		}
+		p.logger.Trace("handle.chunk.delivery", "done put", msg.Addr, "err", err)
+	}()
+	return nil
+}
+
+// RequestFromPeers sends a chunk retrieve request to the next found peer
+func (r *Retrieval) RequestFromPeers(ctx context.Context, req *storage.Request, localID enode.ID) (*enode.ID, error) {
+	log.Debug("retrieval.requestFromPeers", "req.Addr", req.Addr)
+	metrics.GetOrRegisterCounter("network.retrieve.requestfrompeers", nil).Inc(1)
+
+	const maxFindPeerRetries = 5
+	retries := 0
+
+FINDPEER:
+	sp, err := r.findPeer(ctx, req)
+	if err != nil {
+		log.Trace(err.Error())
+		return nil, err
+	}
+
+	protoPeer := r.getPeer(sp.ID())
+	if protoPeer == nil {
+		retries++
+		if retries == maxFindPeerRetries {
+			log.Error("max find peer retries reached", "max retries", maxFindPeerRetries)
+			return nil, ErrNoPeerFound
+		}
+
+		goto FINDPEER
+	}
+
+	ret := RetrieveRequest{
+		Addr: req.Addr,
+	}
+	protoPeer.logger.Trace("sending retrieve request", "ref", ret.Addr, "origin", localID)
+	err = protoPeer.Send(ctx, ret)
+	if err != nil {
+		protoPeer.logger.Error("error sending retrieve request to peer", "err", err)
+		return nil, err
+	}
+
+	spID := protoPeer.ID()
+	return &spID, nil
+}
+
+func (r *Retrieval) Start(server *p2p.Server) error {
```

Comment:
..............?
I'm not sure i'm following you. we agreed to implement retrieval as a separate p2p protocol, that's what i did and p2p protocols need to implement node.Service, that's what I did. I don't understand what you want at this stage. please be kind enough to clarify because it seems you want me to scratch this PR? with which alternative? thanks

Reply:
I am just saying that looking at `Start`, `Stop`, `APIs`,  i dont see a compelling case that `Retrieval` should implement `node.Service` . Just include `Retrieval.Protocol()` among the protocols in `swarm.go`. thats all. But i dont mind